common/src/main/java/com/wynntils/features/ui/BulkBuyFeature.java

```diff
@@ -13,46 +13,161 @@
+
+        // Shop titles are in slot 4, eg. §aScroll Shop
+        // Shops are all size 54 for double chest, sometimes size 41 is sent (no idea what it's for)
+        if (e.getSlot() != 4 || e.getContainer().getContainerSize() != 54) return;
+
+        initBulkBuyWidget(McUtils.mc().screen);
+    }
+
+    @SubscribeEvent
+    public void onContainerOpened(ScreenInitEvent.Post e) {
+        // This event handles any reinit required, eg window resize
+        // It will not make a widget on first init
+        initBulkBuyWidget(e.getScreen());
+    }
+
+    private void initBulkBuyWidget(Screen screen) {
+        // Neither event guarantees that they do not re-fire on the same screen
+        if (bulkBoughtSlotNumber != -1) return;
+
+        if (!(screen instanceof ContainerScreen containerScreen)) return;
+        if (!(containerScreen.getMenu() instanceof AbstractContainerMenu acm)
+                || acm.getItems().size() != 90) return;
+
+        String title = acm.getSlot(4).getItem().getHoverName().getString();
+        if (!title.startsWith(ChatFormatting.GREEN.toString()) || !title.endsWith(SHOP_TITLE_SUFFIX)) return;
+
+        bulkBuyWidget = new BulkBuyWidget(
+                containerScreen.leftPos - Texture.BULK_BUY_PANEL.width(),
+                containerScreen.topPos - 5,
+                Texture.BULK_BUY_PANEL.width(),
+                Texture.BULK_BUY_PANEL.height());
+        containerScreen.addRenderableWidget(bulkBuyWidget);
+    }
+
     @SubscribeEvent
     public void onSlotClicked(ContainerClickEvent e) {
-        if (!KeyboardUtils.isShiftDown()) return;
+        if (!KeyboardUtils.isShiftDown() || e.getClickType() != ClickType.QUICK_MOVE) return;
 
         AbstractContainerMenu container = e.getContainerMenu();
+        ItemStack itemStack = e.getItemStack();
+        if (!isBulkBuyable(container, itemStack)) return;
 
-        if (!isBulkBuyable(container, e.getItemStack())) return;
+        int itemPrice = findItemPrice(LoreUtils.getLore(itemStack));
+        if (itemPrice * (bulkBoughtAmount + bulkBuyAmount.get()) > Models.Emerald.getAmountInInventory()) {
+            McUtils.sendErrorToClient(I18n.get("feature.wynntils.bulkBuy.bulkBuyCannotAfford"));
+            return;
+        }
 
-        if (e.getClickType() == ClickType.QUICK_MOVE) { // Shift + Left Click
-            for (int i = 1; i < bulkBuyAmount.get(); i++) {
-                ContainerUtils.clickOnSlot(e.getSlotNum(), container.containerId, 10, container.getItems());
-            }
+        if (bulkBoughtSlotNumber == -1) {
+            // we're starting a new bulk buy
+            bulkBoughtSlotNumber = e.getSlotNum();
+            bulkBoughtContainer = container;
+            bulkBoughtAmount = bulkBuyAmount.get();
+            bulkBoughtPrice = itemPrice;
+
+            bulkBuyWidget.setBulkBoughtPrice(bulkBoughtPrice);
+            bulkBuyWidget.setBulkBoughtItemStack(itemStack);
```

Comment:
Maybe introduce a `BulkBuyInfo` record to send all of the data at once?

Reply:
I did originally have something like this, but removed it since there are only two instances where I send the price and itemstack together. Do you still think it is worth it?

Reply:
I do think it's worth it, it makes sure you can't missmatch and forget to update part of the data.